src/SignalR/common/Shared/MessageBuffer.cs

```diff
@@ -181,19 +180,11 @@ public void Ack(AckMessage ackMessage)
 
     internal bool ShouldProcessMessage(HubMessage message)
     {
-        // TODO: if we're expecting a sequence message but get here should we error or ignore or maybe even continue to process them?
-        if (_waitForSequenceMessage)
```

Comment:
> While our client implementations shouldn't send messages on reconnect before the `SequenceMessage`, it doesn't hurt to allow that.

So now that this check is no longer here, shouldn't we log something if we receive a `SequenceMessage` after any other message? I don't see how allowing it helps any scenario. I think we should probably abort the connection and raise the clearest error we can.

That probably ends up being an info log since it can be client induced, but we should probably do something, right? It seems like a clear bug that could indicate the client is sending duplicate or out of order messages.

Reply:
There are a few reasons I can think of for a client sending a message before the `SequenceMessage`.

The first one that might happen today is a `PingMessage`. The clients currently don't stop the ping loop so it's possible to reconnect and send the ping message immediately.

The other two I can think of are possible future enhancements. If the client doesn't have any buffered messages (everything was acked before disconnect) it could just not send a sequence message and pretend like the connection never went down. Or we could consider sending an Ack on reconnect so the `SequenceMessage` and subsequent replayed messages will only contain what hasn't been received by either side.

All three of these would work if we don't protect against messages before `SequenceMessage`.

Reply:
> However, I mostly want to detect if we receive any HubInvocationMessages before a SequenceMessage. I know clients shouldn't do that, but that could lead to some weird behavior, right?

I see, if a client potentially buffers multiple messages during the disconnection window, and on reconnect sent the latest message before the `SequenceMessage`, then the server and client would be out of sync by 1 message. If we were to not allow any messages except the `SequenceMessage` that would avoid this issue. It does feel like an edge case that we could clean up in the future.

> I think we need to update the IStatefulReconnectFeature design to make it easier to detect the boundaries between data received over the old and new connections. I know we were talking about removing the PipeWriter argument from the OnReconnected callback, and I'm thinking we should do that by making it an IDuplexPipe instead. Or maybe a whole new ConnectionContext.

Yeah, we've talked about this 😃 
David wants to wrap the Pipes and have them swap internally so the Pipe usage code wouldn't need to change at all.

> Right now, if a client reconnects prior the server calling OnReconnected, it will just miss the reconnection, right?

Yeah, I think it also won't be reading or writing to the Pipe yet, so will just start reading/writing as if the new connection was always the previous connection.